pkg/app/ops/firestoreindexensurer/gcloud.go

```diff
@@ -92,19 +94,59 @@ func (c *gcloud) createIndex(ctx context.Context, idx *index) error {
 	}
 
 	c.logger.Info("start creating a Firestore index", zap.Strings("command", args))
-	if err := c.runGcloudCommand(ctx, args...); err != nil {
+	if _, err := c.runGcloudCommand(ctx, args...); err != nil {
 		return err
 	}
 	return nil
 }
 
-func (c *gcloud) runGcloudCommand(ctx context.Context, args ...string) error {
-	var stderr bytes.Buffer
+func (c *gcloud) listIndexes(ctx context.Context) ([]index, error) {
+	type rawIndex struct {
+		Name       string  `json:"name"`
+		QueryScope string  `json:"queryScope"`
+		Fields     []field `json:"fields"`
+	}
+	args := []string{
+		"firestore", "indexes", "composite", "list",
+		"--sort-by", "name",
+		"--format", "json",
+		"--project", c.projectID,
+	}
+	out, err := c.runGcloudCommand(ctx, args...)
+	if err != nil {
+		return nil, err
+	}
+	var rawIndexes []rawIndex
+	if err := json.Unmarshal(out, &rawIndexes); err != nil {
+		return nil, fmt.Errorf("failed to parse indexes list: %w", err)
+	}
+
+	// Start converting the raw indexes into our own index type.
+	indexes := make([]index, 0, len(rawIndexes))
+	for _, idx := range rawIndexes {
+		// Supposed to be like "projects/project-a/databases/(default)/collectionGroups/CollectionA/indexes/CICAgLjRnZMK"
+		name := strings.Split(idx.Name, "/")
+		if len(name) < 6 {
+			c.logger.Warn("index has unexpected name", zap.String("name", idx.Name))
+			continue
+		}
+		indexes = append(indexes, index{
+			CollectionGroup: name[5],
```

Comment:
I am concerned about using only the collection name to handle the index.
What happens if our users have other collections with the same name at other paths?
Instead, Do we have any way to use the full path as the collection ID?

Reply:
Thank you. At first, we can specify only collection groups like `Application`, `Deployment`, etc. (https://cloud.google.com/sdk/gcloud/reference/firestore/indexes/composite/create)
That is, we cannot create indexes for each paths. The only one collection group with the name `Application` can exist.

Reply:
>Do we have any way to use the full path as the collection ID?

Does it mean we change the collection ID in our GCP project for dev to other one? (kind of like PipeCD/Application)

Reply:
> The only one collection group with the name Application can exist.

Actually, many collections can have the same name as sub-collection and we are using that feature too. You can check the dev environment to see.

> Does it mean we change the collection ID in our GCP project for dev to other one? (kind of like PipeCD/Application)

For example, in a single GCP project, we have multiple pipecd environments (dev, prod) (or db for pipecd and db for user's project) and many collections can have the same name.

The collection path will be something like: `/pipecd/dev/Application`.

From gcloud doc:
> --collection-group=COLLECTION_GROUP
ID of the collection group or fully qualified identifier for the collection group. To set the [collection-group] attribute: provide the argument [--collection-group] on the command line.

Not sure what is the format of the "fully qualified identifier". But the path I wanted to say is this "fully qualified identifier".

Reply:
>Actually, many collections can have the same name as sub-collection and we are using that feature too.

You're totally right. But what I'm trying to say is about a "collection group", not a collection. I assume the collection group is unique inside a GCP project. And I assume our `Application` collection belongs to the collection group named `Application`.

Maybe I got wrong. It might be easy to talk via Zoom 😄